### PR TITLE
C++ SS (CPU part)

### PR DESCRIPTION
**Pull request by orliac**
Porting the standard synthesizer to C++
(Note: to be reviewed by Simon but not available from the Reviewers pane)

### CODE REVIEW

File: src/bluebild/src/host/standard_synthesizer_host.cpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+#include "host/blas_api.hpp"
+#include "memory/allocator.hpp"
+#include "memory/buffer.hpp"
+#include "host/gemmexp.hpp"
+
+
+namespace bluebild {
+
+template <typename T, typename U>
+void cluster_layers(const T*  __restrict__ unlayered_stats,
```
Comment (AdhocMan) [inline src/bluebild/src/host/standard_synthesizer_host.cpp:20]: These functions seem to be used only in this cpp file. So using `static` or an a unnamed namespace would be better.

File: src/bluebild/src/standard_synthesizer.cpp [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+#include "bluebild/bluebild.hpp"
+#include "bluebild/config.h"
+#include "memory/buffer.hpp"
+#include "context_internal.hpp"
+#include "host/standard_synthesizer_host.hpp"
+
+
+namespace bluebild {
+
+    template <typename T>
```
Comment (AdhocMan) [inline src/bluebild/src/standard_synthesizer.cpp:16]: Indentation seems to inconsistent here.

File: src/bluebild/src/standard_synthesizer.cpp [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                              const int Nh,
+                              const int Nw,
+                              T* stats_std,
+                              T* stats_lsq,
+                              T* stats_std_cum,
+                              T* stats_lsq_cum) -> void {
+        auto &ctxInternal = *InternalContextAccessor::get(ctx);
+        if(ctx.processing_unit() == BLUEBILD_PU_GPU) {
+#if defined(BLUEBILD_CUDA) || defined(BLUEBILD_ROCM)
+            //FILL_ME
```
Comment (AdhocMan) [inline src/bluebild/src/standard_synthesizer.cpp:39]: I'd suggest throwing an `NotImplementedError` here.

**Review by AdhocMan — CHANGES_REQUESTED** [review]

File: pypeline/phased_array/bluebild/field_synthesizer/spatial_domain.py [hunk 4]
```diff
@@ -13,6 +13,7 @@
 import scipy.linalg as linalg
 import scipy.sparse as sparse
 import sys
+import nvtx
```
[the remark this answers is not in the document]
Comment (orliac) [reply pypeline/phased_array/bluebild/field_synthesizer/spatial_domain.py]: Removed for now

File: src/bluebild/CMakeLists.txt [hunk 5]
```diff
@@ -14,12 +14,17 @@ if(POLICY CMP0022)
 	cmake_policy(SET CMP0022 NEW)
 endif()
 
-# set default build type to RELEASE
+# set default build type to RELEASE if not defined in corresponding environment variable
+if(DEFINED ENV{CMAKE_BUILD_TYPE})
```
[the remark this answers is not in the document]
Comment (orliac) [reply src/bluebild/CMakeLists.txt]: I now rely on -DCMAKE_BUILD_TYPE and not on environment variables anymore to tune the compilation options
(Note: we are on CMake 3.16 at EPFL)

File: src/bluebild/CMakeLists.txt [hunk 6]
```diff
@@ -32,7 +37,7 @@ include(GNUInstallDirs)
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${PROJECT_SOURCE_DIR}/cmake/modules)
 
 # Options
-option(BLUEBILD_BUILD_TESTS "Build tests" OFF)
+option(BLUEBILD_BUILD_TESTS "Build tests" ON)
```
[the remark this answers is not in the document]
Comment (orliac) [reply src/bluebild/CMakeLists.txt]: Leftover from testing, set back to OFF

File: src/bluebild/CMakeLists.txt [hunk 7]
```diff
@@ -151,13 +157,19 @@ configure_file(include/bluebild/config.h.in ${PROJECT_BINARY_DIR}/bluebild/confi
 list(APPEND BLUEBILD_INCLUDE_DIRS ${PROJECT_SOURCE_DIR}/src)
 list(APPEND BLUEBILD_INCLUDE_DIRS ${PROJECT_SOURCE_DIR}/include)
 list(APPEND BLUEBILD_INCLUDE_DIRS ${PROJECT_BINARY_DIR})
+if(DEFINED ENV{MARLA_ROOT})
```
[the remark this answers is not in the document]
Comment (orliac) [reply src/bluebild/CMakeLists.txt]: I've added a FindMarla.cmake package

File: src/bluebild/include/bluebild/bluebild.hpp [hunk 8]
```diff
@@ -33,4 +33,16 @@ intensity_field_data(Context &ctx, T wl, int m, int n, int nEig,
                      T *d, std::complex<T> *v, int ldv, int nCluster,
                      const T *cluster, int *clusterIndices) -> void;
 
+    //template <typename T, typename = std::enable_if_t<std::is_same_v<T, double> ||
```
[the remark this answers is not in the document]
Comment (orliac) [reply src/bluebild/include/bluebild/bluebild.hpp]: Type restriction back in. I had to fix an issue to make it work (a missing template argument in pybluebild.cpp)

File: src/bluebild/python/pybluebild.cpp [hunk 9]
```diff
@@ -45,6 +45,35 @@ auto check_2d_array(const py::array_t<T, STYLE> &a,
...
+                               const py::array_t<T, py::array::f_style> &grid,
+                               const T wl,
+                               const long Na, const long Nb, const long Nc, const long Ne,
+                               const long Nh, const long Nw,
+                               py::array_t<T, py::array::f_style> &stats_std,
+                               py::array_t<T, py::array::f_style> &stats_lsq,
+                               py::array_t<T, py::array::f_style> &stats_std_cum,
+                               py::array_t<T, py::array::f_style> &stats_lsq_cum
+                               ) {
+    standard_synthesizer(ctx,
```
[the remark this answers is not in the document]
Comment (orliac) [reply src/bluebild/python/pybluebild.cpp:86]: All inputs are now checked for dimensions

File: src/bluebild/src/CMakeLists.txt [hunk 10]
```diff
@@ -67,6 +70,13 @@ endmacro()
 bluebild_create_library(bluebild)
 set_target_properties(bluebild PROPERTIES VISIBILITY_INLINES_HIDDEN TRUE CXX_VISIBILITY_PRESET hidden)
 
+
+# OpenMP
+find_package(OpenMP REQUIRED)
+if(OpenMP_CXX_FOUND)
+    target_link_libraries(bluebild PUBLIC OpenMP::OpenMP_CXX)
```
[the remark this answers is not in the document]
Comment (orliac) [reply src/bluebild/src/CMakeLists.txt]: Done

File: src/bluebild/src/bluebild_export.h [hunk 11]
```diff
@@ -0,0 +1,42 @@
+
```
[the remark this answers is not in the document]
Comment (orliac) [reply src/bluebild/src/bluebild_export.h]: Sorry, this was a mistake, I removed the file

File: src/bluebild/src/host/standard_synthesizer_host.cpp (hunk 1 above)
Comment (orliac) [reply src/bluebild/src/host/standard_synthesizer_host.cpp:21]: Done

File: src/bluebild/src/standard_synthesizer.cpp (hunk 2 above)
Comment (orliac) [reply src/bluebild/src/standard_synthesizer.cpp]: Indentation corrected

**Review by AdhocMan — APPROVED** [review]
Looks good to me now.
Only one suggestion:
Store the test data in a compact json format (without line breaks and spaces), to keep the repository size in check. Definitively starting to notice long checkout times and you can at least half the size of these json files. Maybe at some point, we will have to think about a compressed file format.

**Pull request closed without merging**